### PR TITLE
fix(html_formatter): preserve whitespace around inline elements and text expressions

### LINKED ISSUE #7875: 📝 Spaces are incorrectly removed after elements and expressions in HTML-ish languages
### Environment information

```bash
CLI:
  Version:                      2.3.0
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.21.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Formatter:
  Format with errors:           unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   99
  Attribute position:           unset
  Bracket spacing:              unset
  Includes:                     unset

JavaScript Formatter:
  Enabled:                      unset
  JSX quote style:              unset
  Quote properties:             unset
  Trailing commas:              unset
  Semicolons:                   unset
  Arrow parentheses:            unset
  Bracket spacing:              unset
  Bracket same line:            unset
  Quote style:                  unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           unset

JSON Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Trailing Commas:              unset
  Expand lists:                 unset

CSS Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Quote style:                  unset

GraphQL Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Bracket spacing:              unset
  Quote style:                  unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
  "$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
  "html": {
    "experimentalFullSupportEnabled": true
  }
}
```

### Playground link

https://biomejs.dev/playground/?code=PABkAGkAdgA%2BAAoAIAAgADwAcAA%2BAEkAJwBtACAAagB1AHMAdAAgADwAcwBwAGEAbgA%2BAHQAZQBzAHQAaQBuAGcAPAAvAHMAcABhAG4APgAgAHMAbwBtAGUAdABoAGkAbgBnADwALwBwAD4ACgAgACAAPABwAD4AewBuAGEAbQBlAH0AIAAoAHsAdQBzAGUAcgBuAGEAbQBlAH0AKQA8AC8AcAA%2BAAoAPAAvAGQAaQB2AD4ACgA%3D&language=astro

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

### LINKED ISSUE #7960: 🐛 Unexpected formatting of whitespace in HTML after inline elements within text
### Environment information

```block
CLI:
  Version:                      2.3.2
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-ghostty
  JS_RUNTIME_VERSION:           v22.12.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.6.5

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Workspace:
  Open Documents:               0
```

### What happened?

1.  Unexpected result of formatting space in `html` / `html-ish` files. Whitespace is removed after inline elements within a `<p>` tag when `"whitespaceSensitivity": "strict"` is enabled.

[Playground](https://biomejs.dev/playground/?attributePosition=multiline&whitespaceSensitivity=strict&indentScriptAndStyle=true&code=PAAhAC0ALQAgAEUAeABwAGUAYwB0AGUAZAAgAGYAbwByAG0AYQB0AHQAaQBuAGcAIAB3AGkAdABoACAAYABzAHQAcgBpAGMAdABgACAAdwBoAGkAdABlAHMAcABhAGMAZQBTAGUAbgBzAGkAdABpAHYAaQB0AHkAOgAKACAAYAA8AC8AcwBwAGEAbgA%2BACAAYQBkAGkAcABpAHMAaQBjAGkAbgBnACAAYAAKAC0ALQA%2BAAoAPABwAD4ACgAJAEwAbwByAGUAbQAsACAAaQBwAHMAdQBtACAAZABvAGwAbwByACAAcwBpAHQAIABhAG0AZQB0ACAAPABzAHAAYQBuAD4AYwBvAG4AcwBlAGMAdABlAHQAdQByADwALwBzAHAAYQBuAD4AIABhAGQAaQBwAGkAcwBpAGMAaQBuAGcAIABlAGwAaQB0AC4ACgA8AC8AcAA%2BAAoACgA8ACEALQAtAAoACQBNAG8AdgBpAG4AZwAgAHMAdQBiAHMAZQBxAHUAZQBuAHQAIAB0AGUAeAB0ACAAdABvACAAbgBlAHcAbABpAG4AZQAgAHAAcgBlAHMAZQByAHYAZQBzACAAdwBoAGkAdABlAHMAcABhAGMAZQAuAAoALQAtAD4ACgA8AHAAPgAKAAkATABvAHIAZQBtACwAIABpAHAAcwB1AG0AIABkAG8AbABvAHIAIABzAGkAdAAgAGEAbQBlAHQAIAA8AHMAcABhAG4APgBjAG8AbgBzAGUAYwB0AGUAdAB1AHIAPAAvAHMAcABhAG4APgAKAAkAYQBkAGkAcABpAHMAaQBjAGkAbgBnACAAZQBsAGkAdAAuAAoAPAAvAHAAPgA%3D&language=html&ruleDomains.project=recommended)

```html
<p>
	Lorem, ipsum dolor sit amet <span>consectetur</span> adipisicing elit.
</p>

<!--
	Moving subsequent text to newline preserves whitespace.
-->
<p>
	Lorem, ipsum dolor sit amet <span>consectetur</span>
	adipisicing elit.
</p>
```

2. The following output is observed for the 1st case above. Whitespace is removed after the closing tag.

```html
<p>
	Lorem, ipsum dolor sit amet <span>consectetur</span>adipisicing elit.
</p>
```

```json
// biome.json

{
  "$schema": "https://biomejs.dev/schemas/2.3.2/schema.json",
  "html": {
    "experimentalFullSupportEnabled": true,
    "formatter": {
      "enabled": true,
      "indentScriptAndStyle": true,
      "whitespaceSensitivity": "strict"
    }
  },
}
```

### Expected result

Expected formatting with `strict` `whitespaceSensitivity`:

```html
<p>
	Lorem, ipsum dolor sit amet <span>consectetur</span> adipisicing elit.
</p>
```

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

### PR DESCRIPTION
**Pull request by matanshavit**
## Summary
Fixes #7875, fixes #7960 where whitespace was being incorrectly removed after closing tags of inline elements and around text expressions (e.g., {name} in Astro/Vue/Svelte). 

## AI assistance notice
I used Claude Code to write most of the code and comments in this pull request. As the human author, I take full responsibility for its contents.

## Test Plan

I added snapshot tests for the straightforward and edge cases around preserving whitespace around text expressions and in tags in whitespace-sensitve mode.

Some snapshots that test other things like long tag children needed to be updated, but they appear to be unrelated to whitespace preservation, and can be safely changed without changing the meaning of the tests.

Note - unimplemented nodes/tokens are expected with snapshots of framework tests because of limited support for html-ish languages (compared to support for just html).

### CODE REVIEW

File: crates/biome_html_formatter/tests/specs/html/text_expressions/expressions_line_boundaries.astro.snap [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+Bracket same line: false
+Whitespace sensitivity: css
+Indent script and style: false
+Self close void elements: never
+-----
+
+```astro
+<p>
+	Start {expr1}
+	{expr2}middle end {expr3}
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/text_expressions/expressions_line_boundaries.astro.snap:41]: The space from the input `{expr2} middle` disappeared in the output.

It's a bit weird that it keeps the newline between `{expr1}` and `{expr2}`, but [it matches prettier behavior](https://biomejs.dev/playground/?lintRules=useSimplifiedLogicExpression&analyzerFixMode=safeAndUnsafeFixes&code=PABwAD4ACgAJAFMAdABhAHIAdAAgAHsAZQB4AHAAcgAxAH0ACgAJAHsAZQB4AHAAcgAyAH0AIABtAGkAZABkAGwAZQAKAAkAZQBuAGQAIAB7AGUAeABwAHIAMwB9AAoAPAAvAHAAPgAKAAoAPABkAGkAdgA%2BAHsAYQB9AAoAewBiAH0APAAvAGQAaQB2AD4A&language=svelte), so keep it.

File: crates/biome_html_formatter/tests/specs/html/text_expressions/multiple_expressions.vue.snap [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+info: text_expressions/multiple_expressions.vue
+---
+# Input
+
+```vue
+<template>
+	<p>{a} {b} {c}</p>
+	<div>{firstName}{middleInitial}.{lastName}</div>
+	<span> {padded} </span>
+</template>
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/text_expressions/multiple_expressions.vue.snap:12]: Vue uses `{{` and `}}`

File: crates/biome_html_formatter/tests/specs/html/text_expressions/single_expressions_whitespace.vue.snap [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+source: crates/biome_formatter_test/src/snapshot_builder.rs
+info: text_expressions/single-expressions-whitespace.vue
+---
+# Input
+
+```vue
+<template>
+	<p>User: {name} ({username})</p>
+	<div>{count} items in cart</div>
+	<span>{first} {last}</span>
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/text_expressions/single_expressions_whitespace.vue.snap:11]: same thing here

File: crates/biome_html_formatter/tests/specs/html/component-frameworks/astro-expressions-whitespace.astro.snap [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+-----
+
+```astro
+<div>
+	<p>
+		I'm just <span>testing</span> something
+	</p>
+	<p>{name}({username})</p>
+	<p>Hello {firstName} {lastName}!</p>
+	<p>{a}+ {b}= {sum}</p>
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/component-frameworks/astro-expressions-whitespace.astro.snap:43]: 1. The space between `{name}` and `({username})` disappeared.
1. The spaces before the `+` and `=` disappeared.

File: crates/biome_html_formatter/tests/specs/html/elements/inline/mixed-expressions-elements.astro.snap [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	{fullName}
+	<span>Age:</span>
+	{age}
+</div>
+
+<p>
+	<strong>Total:</strong>
+	{price}
+	<em>(tax included)</em>
+</p>
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/elements/inline/mixed-expressions-elements.astro.snap:47]: This is significantly different from prettier's output, and shows no difference from biome's current output.

https://biomejs.dev/playground/?code=PABkAGkAdgA%2BAAoACQA8AHMAcABhAG4APgBOAGEAbQBlADoAPAAvAHMAcABhAG4APgAgAHsAZgB1AGwAbABOAGEAbQBlAH0AIAA8AHMAcABhAG4APgBBAGcAZQA6ADwALwBzAHAAYQBuAD4AIAB7AGEAZwBlAH0ACgA8AC8AZABpAHYAPgAKAAoAPABwAD4APABzAHQAcgBvAG4AZwA%2BAFQAbwB0AGEAbAA6ADwALwBzAHQAcgBvAG4AZwA%2BACAAewBwAHIAaQBjAGUAfQAgADwAZQBtAD4AKAB0AGEAeAAgAGkAbgBjAGwAdQBkAGUAZAApADwALwBlAG0APgA8AC8AcAA%2BAA%3D%3D&language=svelte

File: crates/biome_html_formatter/tests/specs/html/text_expressions/expressions_ignore/expressions-whitespace-ignore.astro.snap [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+Line width: 80
+Attribute Position: Auto
+Bracket same line: false
+Whitespace sensitivity: css
+Indent script and style: false
+Self close void elements: never
+-----
+
+```astro
+<p>{name}({username})</p>
```
Comment (dyc3) [inline crates/biome_html_formatter/tests/specs/html/text_expressions/expressions_ignore/expressions-whitespace-ignore.astro.snap:32]: The space disappeared

**Comment by ematipico** [comment]
It would be nice if the PR description would highlight what kind of strategy (business logic) was used to fix the bug. 

Without it, I wouldn't know what to review and how.

File: crates/biome_html_formatter/src/utils/metadata.rs [hunk 7]
```diff
@@ -809,6 +817,14 @@ pub(crate) fn is_inline_element(tag_name: &HtmlTagName) -> bool {
         .any(|tag| tag_name.text_trimmed().eq_ignore_ascii_case(tag))
 }
 
+/// Whether the element is a text expression (like `{name}` in HTML-ish languages)
+pub(crate) fn is_text_expression(element: &AnyHtmlElement) -> bool {
```
Comment (ematipico) [inline crates/biome_html_formatter/src/utils/metadata.rs:821]:
```suggestion
const pub(crate) fn is_text_expression(element: &AnyHtmlElement) -> bool {
```

File: crates/biome_html_formatter/src/html/auxiliary/closing_element.rs [hunk 8]
```diff
@@ -57,6 +57,14 @@ impl FormatNodeRule<HtmlClosingElement> for FormatHtmlClosingElement {
 
         write!(f, [r_angle_token.format()])?;
 
+        // Handle whitespace sensitivity for trailing whitespace after the closing tag's `>`
+        if let Ok(r_angle_token) = &r_angle_token
+            && is_whitespace_sensitive
+            && r_angle_token.has_trailing_whitespace()
+        {
+            write!(f, [space()])?;
```
Comment (ematipico) [inline crates/biome_html_formatter/src/html/auxiliary/closing_element.rs:65]: Why do we add a whitespace?

**Comment by alissonlauffer** [comment]
I see there are other issues related to whitespaces. I think we should revisit these to know whether this PR also resolves them:
#4927
#6407

My issue (#7875) was almost a duplicate of these, but it also encompasses expressions.
